### PR TITLE
Cancel timeout in PersistentFSM

### LINKED ISSUE #22166: `PersistentFSM` misinterprets `Duration.Inf` to disable timeout for state.
Given ```when(MyState, someDuration) {...}```
When using `goto(MyState) forMax Duration.Inf` in a `PersistentFSM`, I was expecting that the StateTimeout event was not received.

It turns out that this is converted to a `None` value for the `timeout` for `State` (see https://github.com/akka/akka/blob/master/akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala#L303).

In `FSM` there seems to be some special logic to deal with this (see https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/FSM.scala#L169). I guess this should be ported to `PersistentFSM` to make this work as the documentation specifies:

```
/**
     * Modify state transition descriptor to include a state timeout for the
     * next state. This timeout overrides any default timeout set for the next
     * state.
     *
     * Use Duration.Inf to deactivate an existing timeout.
     */
```

Related to #17140, but for `PersistentFSM` instead of `FSM`

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #22166

### CODE REVIEW

File: akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala [hunk 1]
```diff
@@ -133,6 +133,13 @@ trait PersistentFSM[S <: FSMState, D, E] extends PersistentActor with Persistent
 }
 
 object PersistentFSM {
+
+  /**
+   * Used by `forMax` to signal "cancel stateTimeout"
+   * INTERNAL API
+   */
+  private[fsm] final val SomeMaxFiniteDuration = Some(Long.MaxValue.nanos)
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala]: `@InternalApi`?

Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala]: (we'll make MiMa understand those annotations soon)

Comment (johanandren) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala]: Oh, didn't see that one before.

**Comment by ktoso** [comment]
Ah right, you were away then.
We now have:
- https://github.com/akka/akka/pull/22111
- and https://github.com/akka/akka-http/issues/438

Originated by need in: https://github.com/akka/akka-http/issues/438

-- 
Konrad `ktoso` Malawski
Akka <http://akka.io> @ Lightbend <http://lightbend.com>

On 20 January 2017 at 11:58:18, Johan Andrén (notifications@github.com)
wrote:

*@johanandren* commented on this pull request.
------------------------------

In akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSM.scala
<https://github.com/akka/akka/pull/22173>:

> @@ -133,6 +133,13 @@ trait PersistentFSM[S <: FSMState, D, E] extends PersistentActor with Persistent
 }

 object PersistentFSM {
+
+  /**
+   * Used by `forMax` to signal "cancel stateTimeout"
+   * INTERNAL API
+   */
+  private[fsm] final val SomeMaxFiniteDuration = Some(Long.MaxValue.nanos)

Oh, didn't see that one before.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/akka/akka/pull/22173>, or mute the thread

File: akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMSpec.scala [hunk 2]
```diff
@@ -329,6 +329,20 @@ abstract class PersistentFSMSpec(config: Config) extends PersistenceSpec(config)
...
+    "allow cancelling stateTimeout by issuing forMax(Duration.Inf)" in {
+      val p = TestProbe()(system)
+
+      val fsm = system.actorOf(TimeoutFSM.props(p.ref))
+
+      p.expectMsg(PersistentFSM.StateTimeout)
+
+      fsm ! TimeoutFSM.OverrideTimeoutToInf
+      p.expectMsg(TimeoutFSM.OverrideTimeoutToInf)
+      p.expectNoMsg(3.seconds)
```
Comment (patriknw) [inline akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMSpec.scala]: 1.second would be enough if the other timeout is shorter (given that we run the tests many times)

File: akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMSpec.scala [hunk 3]
```diff
@@ -506,6 +520,33 @@ object PersistentFSMSpec {
...
+  class TimeoutFSM(probe: ActorRef)(implicit val domainEventClassTag: ClassTag[String]) extends Actor
+    with PersistentFSM[TimeoutFSM.State, String, String] {
+
+    import TimeoutFSM._
+    override def applyEvent(domainEvent: String, currentData: String): String = "whatever"
+    override def persistenceId: String = "timeout-test"
+
+    startWith(State("init"), "")
+
+    when(State("init"), stateTimeout = 1.second) {
```
Comment (patriknw) [inline akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMSpec.scala]: make it shorter to speed up the test

**Review by patriknw — APPROVED** [review]
LGTM, after speedup

**Comment by patriknw** [comment]
thanks

**Pull request merged**